rfcs/00xx-select-from-subscription.md

```diff
@@ -0,0 +1,188 @@
+
+FETCH NEXT FROM cur;
+  pk   |   v   | op | rw_timestamp 
+-------+-------+----+--------------
+ 25184 | 25184 |  1 |             
+(1 row)
+
+FETCH NEXT FROM cur;
+ t.pk  |  t.v  | t.op | rw_timestamp  
+-------+-------+------+---------------
+ 18704 | 18704 |    2 | 1715021447040
+(1 row)
+
+...
+
+```
+
+Basically we support `FETCH NEXT FROM <cursor>` on a `subscription` to get the changelog stream with schema `(original schema | op | rw_timestamp)`.
+
+
+This RFC will extend the usage of `subscription` to support `SELECT` query on it as well. The output schema of the SELECT query is `(original schema | op | rw_timestamp)`.
+
+## Design
+### Batch SELECT on subscription
+```SQL
+CREATE TABLE t (pk int PRIMARY KEY, v int);
+
+INSERT INTO t values (1, 1);
+
+CREATE SUBSCRIPTION sub FROM t WITH (
+    retention = '1D'
+);
+
+
+INSERT INTO t VALUES (2, 2);
+
+UPDATE t SET v = 11 WHERE pk = 1;
+
+# op = 1 insert
+# op = 2 delete
+# op = 3 update_insert
+# op = 4 update_delete
+SELECT * FROM sub;
+ t.pk  |  t.v  | t.op | rw_timestamp  
+-------+-------+------+---------------
+ 2     | 2     |    1 | 1715021447030    
+ 1     | 1     |    4 | 1715021447040
+ 1     | 11    |    3 | 1715021447040
+(3 rows)
+
+SELECT t.pk as pk, t.v as v, CASE WHEN t.op = 2 or t.op = 4 THEN TRUE ELSE FALSE END as is_delete FROM sub;
+ pk    |   v   |   is_delete   
+ ------+-------+------------
+ 2     | 2     |    f     
+ 1     | 1     |    t 
+ 1     | 11    |    f 
+(3 rows)
+```
+
+- Similar to the requirement for CURSOR query, `retention` must be specified when creating a subscription, that is eligble for batch select.
+- Unlike CUSOR query, batch select won't see historical snapshot data when the subscription is created. It only sees the data after the subscription is created. Technically the following two queries are equivalent:
+```SQL
+SELECT * FROM sub;
+
+# equivalent to FETCH NEXT from cursor created with since begin() till rw_timestamp == now()
+
+DECLARE cur SUBSCRIPTION CURSOR FOR sub SINCE BEGIN();
+FETCH NEXT FROM cur;
+...
+```
+
+Use case: one-shot inspection on the MV/Table changelog for debugging
+
+
+### Streaming SELECT on subscription
+```SQL
+CREATE TABLE t (pk int PRIMARY KEY, v int);
+
+CREATE SUBSCRIPTION sub FROM t;
+
+CREATE MATERIALIZED VIEW t_changelog AS
+SELECT * FROM sub;
+
+CREATE SINK t_changelog_sink
+AS SELECT t.pk as pk, t.v as v, CASE WHEN t.op = 2 or t.op = 4 THEN TRUE ELSE FALSE END as is_delete FROM sub;
+```
+
+- User can create sink/mv on the subscription to get the changelog stream of the MV/Table the subscription is created upon.
+- `retention` is optional for subscription eligble for streaming query.
+  - When `retention` is set, the corresponding log store will be created and the subscription is eligble for cursor query, batch select query, and streaming query.
+  - When `retention` is not set, no log store will be created for the subscription and the subscription is only eligble for streaming query.
```

Comment:
Discussion: should we have a default retention like 10mins when retention is not set for convenience to make all subscriptions eligible for cursor query, batch select query and streaming query?

Reply:
In my initial design, `retention` is a mandatory parameter to ensure that users are aware of the data cleaning.
I still believe that it's a right design choice.

Reply:
For the "create sink/mv from subscription" use case, we don't necessarily need to buffer the changelog data under global checkpoint. The reason why I make `retention` optional in this use case is because we can skip creating the log store for subscription and in this case there is no data to clean.